### DISCUSSION TITLE
Why wstring doesnt supported by your json library?

### DISCUSSION THREAD

**Discussion by lipasite** [nlohmann/json#2641] [Q&A] [answered]
I am trying to do this - converting from wstring works incorrectly!
I need add to json array of objects **wstring** values,
and then convert to plain **wchar_t** string!

But all I try - didnt work! This is [screenshot](https://imgur.com/a/mzHiCJI)! 

```
auto jsonProducts = json::array();
while (...) {
  std::wstring productRID_w (ValByName(GoodsQ));
  std::wstring productName_w (ValByName2(GoodsQ));
  // convert to json
  json product = {
	{"productRID", productRID },
	{"productName", productName}
  };
jsonProducts.push_back(product);
}
std::string jStr = jsonProducts.dump();
std::wstring widestr = std::wstring(jStr.begin(), jStr.end());
wchar_t* tempWide = const_cast<wchar_t*>(widestr.c_str());
return tempWide;
```

So, the wstring **L"test"** after adding to jon object becomes **"Р\xfҐP\a"** 

What the hell, man?
How to fix it?
Or suggest a normal C ++ library that can work correctly with wstring.
Be so kind....

**Comment by nlohmann** [comment, marked as answer]
This library does not support `std::wstring`, but only `std::string` with UTF-8 encoding.

**Reply by NN---** [reply]
Please consider adding wstring support.
There are many developers who cannot use this library due to lack of wstring support.

**Reply by falbrechtskirchinger** [reply]
It might not be too difficult to add conversion to and from `std::wstring` to the library using https://en.cppreference.com/w/cpp/locale/codecvt. Meanwhile, you'll have to do the conversion yourself, which isn't too difficult either.

**Reply by NN---** [reply]
It is not hard to convert.
I talk about cases where you receive UTF-16 JSON and return UTF-16 JSON.
In such case a conversion is totally unnecessary.

You can say it is an old legacy to use UTF-16 and you will be correct.
There is still a usage for this in Windows or other old systems.

**Reply by falbrechtskirchinger** [reply]
The library can parse wide strings but the end result is always stored as a narrow string. That is unlikely to change as a lot of code depends on the assumption that the char type is always 1 byte in size.
